src/charm.py

```diff
@@ -316,8 +326,16 @@ def _handle_auth_config(self, _):
         """Event handler for forward_auth config changes."""
         self._sync_all_resources()
 
+    def _on_peers_changed(self, _):
+        """Event handler for whenever peer topology changes."""
+        self._sync_all_resources()
+
     def _on_remove(self, _):
         """Event handler for remove."""
+        # if there are still units left, skip removal
+        if self.model.app.planned_units() > 0:
+            return
```

Comment:
Since this is novel and we might want to debug it in future, lets add some helpful logging

```suggestion
        if self.model.app.planned_units() > 0:
            logger.info("Handling remove event: skipping resource removal because application is not scaling to 0.")
            return

        logger.info(
            "Handling remove event: Attempting to remove application resources because application is scaling to 0."
        )

```